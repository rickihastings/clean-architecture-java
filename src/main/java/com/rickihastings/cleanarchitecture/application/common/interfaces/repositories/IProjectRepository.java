package com.rickihastings.cleanarchitecture.application.common.interfaces.repositories;

import com.rickihastings.cleanarchitecture.domain.entities.Project;
import org.springframework.stereotype.Repository;
import java.util.List;
import java.util.Optional;

@Repository
public interface IProjectRepository {
    List<Project> findAll();
    Optional<Project> findById(Long id);
    Project save(Project project);
}
