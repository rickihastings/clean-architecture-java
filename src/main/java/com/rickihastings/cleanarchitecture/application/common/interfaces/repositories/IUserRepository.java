package com.rickihastings.cleanarchitecture.application.common.interfaces.repositories;

import com.rickihastings.cleanarchitecture.domain.User;
import org.springframework.stereotype.Repository;

import java.util.Optional;

@Repository
public interface IUserRepository {
    Optional<User> getUserByUsername(String username);
    User save(User save);
}
