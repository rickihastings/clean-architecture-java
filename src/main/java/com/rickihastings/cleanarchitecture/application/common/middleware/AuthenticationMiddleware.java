package com.rickihastings.cleanarchitecture.application.common.middleware;

import an.awesome.pipelinr.Command;
import com.rickihastings.cleanarchitecture.application.common.annotations.Authenticate;
import com.rickihastings.cleanarchitecture.application.common.exceptions.UnauthorizedException;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;

@Component
public class AuthenticationMiddleware implements Command.Middleware {

    @Override
    public <R, C extends Command<R>> R invoke(C command, Next<R> next) {
        var hasAuthenticate = command.getClass().isAnnotationPresent(Authenticate.class);

        if (hasAuthenticate) {
            var auth = SecurityContextHolder.getContext().getAuthentication();

            if (auth != null && auth.getAuthorities().stream().anyMatch(a -> a.getAuthority().equals("ROLE_USER"))) {
                return next.invoke();
            }

            throw new UnauthorizedException();
        }

        return next.invoke();
    }
}
