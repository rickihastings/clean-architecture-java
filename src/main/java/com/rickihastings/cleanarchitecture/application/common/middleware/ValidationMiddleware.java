package com.rickihastings.cleanarchitecture.application.common.middleware;

import an.awesome.pipelinr.Command;
import com.rickihastings.cleanarchitecture.application.common.exceptions.ValidationException;
import org.springframework.stereotype.Component;
import javax.validation.ConstraintViolation;
import javax.validation.Validator;
import javax.validation.ValidatorFactory;
import java.util.ArrayList;
import java.util.Collection;
import java.util.Set;

@Component
public class ValidationMiddleware implements Command.Middleware {

    private final Validator validator;

    public ValidationMiddleware(ValidatorFactory factory) {
        this.validator = factory.getValidator();
    }

    @Override
    public <R, C extends Command<R>> R invoke(C command, Next<R> next) {
        Set<ConstraintViolation<C>> violations = validator.validate(command);

        if (violations.size() > 0) {
            Collection<Throwable> exceptions = new ArrayList<>();
            for (ConstraintViolation<C> violation : violations) {
                exceptions.add(new Exception(violation.getMessage()));
            }

            throw new ValidationException(exceptions);
        }

        return next.invoke();
    }
}
