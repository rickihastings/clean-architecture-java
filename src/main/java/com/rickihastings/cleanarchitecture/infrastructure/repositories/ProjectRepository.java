package com.rickihastings.cleanarchitecture.infrastructure.repositories;

import com.rickihastings.cleanarchitecture.application.common.interfaces.repositories.IProjectRepository;
import com.rickihastings.cleanarchitecture.domain.Project;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.lang.NonNull;

import java.util.List;

public interface ProjectRepository extends IProjectRepository, JpaRepository<Project, Long> {
    List<Project> findByArchivedFalse();
    Project save(@NonNull Project project);
}

