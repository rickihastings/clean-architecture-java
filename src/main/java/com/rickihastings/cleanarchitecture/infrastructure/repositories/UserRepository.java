package com.rickihastings.cleanarchitecture.infrastructure.repositories;

import com.rickihastings.cleanarchitecture.application.common.interfaces.repositories.IUserRepository;
import com.rickihastings.cleanarchitecture.domain.entities.User;
import org.jetbrains.annotations.NotNull;
import org.springframework.data.jpa.repository.JpaRepository;
import org.springframework.lang.NonNull;

import java.util.Optional;

public interface UserRepository extends IUserRepository, JpaRepository<User, Long> {
     Optional<User> getUserByUsername(String username);

     @NotNull
     User save(@NonNull User user);
}
