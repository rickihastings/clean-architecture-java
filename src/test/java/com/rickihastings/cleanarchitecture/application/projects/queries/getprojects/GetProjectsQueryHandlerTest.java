package com.rickihastings.cleanarchitecture.application.projects.queries.getprojects;

import an.awesome.pipelinr.Pipelinr;
import au.com.origin.snapshots.Expect;
import au.com.origin.snapshots.junit5.SnapshotExtension;
import com.rickihastings.cleanarchitecture.application.common.interfaces.repositories.IProjectRepository;
import com.rickihastings.cleanarchitecture.application.common.middleware.ValidationMiddleware;
import com.rickihastings.cleanarchitecture.seeds.ProjectSeeds;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.Mockito;

import javax.validation.Validation;
import java.util.stream.Stream;

import static org.mockito.Mockito.when;

@ExtendWith({ SnapshotExtension.class })
public class GetProjectsQueryHandlerTest {

    private Expect expect;

    private final IProjectRepository projectRepository = Mockito.mock(IProjectRepository.class);

    private Pipelinr pipeline;

    @BeforeEach
    void setup() {
        var factory = Validation.buildDefaultValidatorFactory();

        pipeline = new Pipelinr()
                .with(() -> Stream.of(new GetProjectsQueryHandler(projectRepository)))
                .with(() -> Stream.of(new ValidationMiddleware(factory)));
    }

    @Test
    public void shouldReturnProjectsWhenValid() {
        when(projectRepository.findByArchivedFalse()).thenReturn(ProjectSeeds.getProjects());

        var result = pipeline.send(new GetProjectsQuery());

        expect.serializer("json").toMatchSnapshot(result);
    }
}
