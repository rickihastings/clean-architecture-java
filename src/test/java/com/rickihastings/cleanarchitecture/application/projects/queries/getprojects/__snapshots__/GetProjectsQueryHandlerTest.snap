com.rickihastings.cleanarchitecture.application.projects.queries.getprojects.GetProjectsQueryHandlerTest.shouldReturnProjectsWhenValid=[
  [
    {
      "title": "Initial Project",
      "createdAt": "2007-12-03T09:15:30Z",
      "createdBy": "Joe Bloggs"
    },
    {
      "title": "Second Project",
      "createdAt": "2007-12-03T09:15:30Z",
      "createdBy": "Joe Bloggs"
    }
  ]
]